x-pack/dockerlogbeat/pipelineManager/pipelineManager.go

```diff
@@ -0,0 +1,233 @@
+	if err != nil {
+		return nil, err
+	}
+	pm.clients[file] = cl
+	pm.pipelines[hashstring].refCount++
+	// before we finish, prune the client list async
+
+	return cl, nil
+}
+
+// a wrapper for various public functions to create pipelines
+// assumes we have a mutex lock. If the pipeline exists, this does nothing.
+func (pm *PipelineManager) makePipeline(logOptsConfig map[string]string, name, hashstring string) error {
+	_, test := pm.pipelines[hashstring]
+	if test {
+		return nil
+	}
+
+	pipeline, err := loadNewPipeline(logOptsConfig, name, pm.Logger)
+	if err != nil {
+		return errors.Wrap(err, "error loading pipeline")
+	}
+	pm.pipelines[hashstring] = &Pipeline{pipeline: pipeline, refCount: 0}
+	return nil
+}
+
+// makeConfigHash is the helper function that turns a user config into a hash
+func makeConfigHash(cfg map[string]string) string {
+	var hashString string
+	var orderedVal []string
+
+	for _, val := range cfg {
+		orderedVal = append(orderedVal, val)
+	}
+
+	sort.Strings(orderedVal)
+
+	for _, val := range orderedVal {
+		hashString = hashString + val
+	}
+
+	sum := sha1.Sum([]byte(hashString))
+
+	return string(sum[:])
+}
+
+// load pipeline starts up a new pipeline with the given config
+func loadNewPipeline(logOptsConfig map[string]string, name string, log *logp.Logger) (*pipeline.Pipeline, error) {
+	info := beat.Info{
+		Beat:     "dockerlogbeat",
+		Version:  "0",
+		Name:     name,
+		Hostname: "dockerbeat.test",
+	}
+
+	newCfg, err := parseCfgKeys(logOptsConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "error parsing config keys")
+	}
+
+	cfg, err := common.NewConfigFrom(newCfg)
+	if err != nil {
+		return nil, err
+	}
+	config := containerConfig{}
+	err = cfg.Unpack(&config)
+	if err != nil {
+		return nil, fmt.Errorf("unpacking config failed: %v", err)
+	}
+
+	processing, err := processing.MakeDefaultSupport(false)(info, log, cfg)
+	if err != nil {
+		return nil, errors.Wrap(err, "error in MakeDefaultSupport")
+	}
+
+	pipelineCfg := pipeline.Config{}
+	err = cfg.Unpack(&pipelineCfg)
+	if err != nil {
+		return nil, errors.Wrap(err, "error unpacking pipeline config")
+	}
+
+	idx, err := idxmgmt.DefaultSupport(log, info, config.Output.Config())
+	if err != nil {
+		return nil, errors.Wrap(err, "error making index manager")
+	}
+
+	pipeline, err := pipeline.Load(info,
+		pipeline.Monitors{
+			Metrics:   nil,
+			Telemetry: nil,
+			Logger:    log,
+		},
+		pipelineCfg,
+		processing,
+		func(stat outputs.Observer) (string, outputs.Group, error) {
+			cfg := config.Output
+			out, err := outputs.Load(idx, info, stat, cfg.Name(), cfg.Config())
+			return cfg.Name(), out, err
+		},
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "error in pipeline.Load")
+	}
+
+	return pipeline, nil
+}
+
+// parseCfgKeys helpfully parses the values in the map, so users can specify yml structures.
+func parseCfgKeys(cfg map[string]string) (map[string]interface{}, error) {
```

Comment:
If you want similar behavior to other Beats, consider `common.NewConfigFrom(...)`

Reply:
`common.NewConfigFrom(...)` is called after `parseCfgKeys`. I had to hack this in because `NewConfigFrom` doesn't like the `map[string]string` structures we get from docker. I had to coerce the string value into an an interface wrapper so it'll handle config keys like `["192.168.1.2","192.168.1.3"]`

Reply:
Hm, sounds like a bug in `go-ucfg`. It should happily accept map[string]string.

Reply:
Oh, it'll accept `map[string]string,` but if you pass it something like

```go 
map[string]string{
    "output.elasticsearch.hosts" : `["192.168.1.2","192.168.1.3"]`
}
```
You get `Elasticsearch url: http://['172.18.0.3:9200,'172.18.0.2']:9200`

Reply:
oh, I see. The `-E` flag is implemented in `libbeat/cfgfile/cfgfile.go`. It uses `SettingFlag` from `libbeat/common/flags.go`. The final flag parsing is implemented here: https://github.com/elastic/go-ucfg/blob/master/flag/value.go#L48

The value parser is an internal package, but maybe we can export it for use-cases like this: CLI settings getting passed indirectly.